main/src/ca/uwaterloo/flix/util/StringCursor.scala

```diff
@@ -0,0 +1,124 @@
+package ca.uwaterloo.flix.util
+
+/**
+  * A class to iterate through an array of characters while maintaining the line and column index
+  * of the cursor.
+  */
+class StringCursor(val data: Array[Char]) {
+
+  /** The cursor pointing into `data`. */
+  private var offset: Int = 0
+
+  /** The line index of `offset`. */
+  private var line: Int = 0
+
+  /** The column index of `offset` */
+  private var column: Int = 0
+
+  /** The max column index of the previous line or `0` if there is no previous line. */
+  private var prevLineMaxColumn = 0
+
+  /** Returns the current line index. */
+  def getLine: Int = line
+
+  /** Returns the current column index. */
+  def getColumn: Int = column
+
+  /** Returns the current source offset. */
+  def getOffset: Int = offset
+
+  /** Returns `(line, column)`. */
+  def getPosition: (Int, Int) = (line, column)
+
+  /**
+    * Returns `(line, column)` where non-existent positions are preferred instead of the first
+    * position of the next line.
+    *
+    * In this example, the current position is on 'v' (1,0) but this function will then return
+    * the position just after 'Example' (0, 7). This is sometimes preferable for exclusive
+    * end positions of ranges.
+    *
+    * {{{
+    *   Example
+    *   v
+    * }}}
+    */
+  def getExclusiveEndPosition: (Int, Int) = {
+    if (line <= 0 || column > 0) {
+      (line, column)
+    } else {
+      (line - 1, prevLineMaxColumn + 1)
+    }
+  }
+
+  /** Returns a copy of `this`, pointing to the same underlying array. */
+  def copy(): StringCursor = {
```

Comment:
It seems the order of functions is:

getter things
advance/peek things

so I would move the copy function last.